projects/meta_cl/networks/dendritic_networks.py

```diff
@@ -107,3 +109,107 @@ def forward(self, x):
         pred = self.prediction(x)
         out = self.gating_layer(pred, context=mod)
         return out
+
+
+class DendriticNetwork2(nn.Module):
+    """
+    Prototype of a dendritic network, based closely on `ANML`_. (The conv + max-pool
+    layers are identical). One major difference, however, is that no batchnorm
+    layers are included.
+
+    .. _ANML: nupic.research/projects/meta_cl/networks/anml_networks.py
+
+    Input shape expected is 3 x 28 x 28 (C x W x H) but an adaptive average pooling
+    layer helps accept other heights and width.
+
+    With default parameters and `num_classes-963`, it uses 5,118,291 weights-on
+    out of a total of 8,345,095 weights. In comparison, ANML uses 5,963,139 weights
+    and OML uses 5,172,675 weights. Thus, these is still room to add more.
+    """
+
+    def __init__(self, num_classes,
+                 num_segments=20,
+                 dim_context=100,
+                 dendrite_sparsity=0.70,
```

Comment:
In case the performance of this model is really bad, I think it may be because if 70% of dendrite weights are initialized as zero, than they won't get updated in the backward pass, so only 30% of dendrite weights can be useful during the lifetime of the network. That's just a guess - what do you think?

Reply:
Actually what I stated above may be incorrect, since the derivative w.r.t. the dendrite weight just involves the dendrite activation value, and not the zero weight.

Reply:
You're right, the derivative may not be zero. This has a complicated story though.

I meant to keep the weights always zero. The module `DendriteSegments` is supposed to behave very similarly to `SparseWeights`. Both inherit from `HasRezeroWeight` so they both have a `rezero_weights` method that should be called after every gradient step.

However, I just realized I fell short of this. I intended to have the `RezeroWeights` mixin take care of rezero-ing the weights (in PR #416, this mixin now operates over every `HasRezeroWeights` module`), but I did not realize it does so in the `post_batch` method, which `MetaContinualLearningExperiment` doesn't have.

Here's a Jira to accomplish this https://ntaresearch.atlassian.net/browse/RES-1861. I'll take care of this soon, either today or tomorrow so we know the modules are behaving as we want them to.